airbyte-cdk/python/airbyte_cdk/sources/source.py

```diff
@@ -39,15 +38,22 @@ def discover(self, logger: logging.Logger, config: TConfig) -> AirbyteCatalog:
         """
 
 
-class Source(DefaultConnectorMixin, BaseSource[Mapping[str, Any], MutableMapping[str, Any], ConfiguredAirbyteCatalog], ABC):
+class Source(
+    DefaultConnectorMixin,
+    BaseSource[Mapping[str, Any], Union[List[AirbyteStateMessage], MutableMapping[str, Any]], ConfiguredAirbyteCatalog],
+    ABC,
+):
     # can be overridden to change an input state
-    def read_state(self, state_path: str) -> Dict[str, Any]:
+    def read_state(self, state_path: str) -> List[AirbyteStateMessage]:
         if state_path:
             state_obj = json.loads(open(state_path, "r").read())
-        else:
-            state_obj = {}
-        state = defaultdict(dict, state_obj)
-        return state
+            if not state_obj:
+                return []
+            elif isinstance(state_obj, List):
+                return [AirbyteStateMessage.parse_obj(state) for state in state_obj]
```

Comment:
as part of instantiation, pydantic will verify correctness of input. granted a lot of input is optional at the top level state (to support backwards compatability)

Reply:
notes from over the shoulder review: see if pydantic can validate at least 1 of the 3 fields is populated, or add validation here to ensure at least one of global_ or streams is populated.

Reply:
I believe this is what you ended up doing with `Type.check_at_least_one` . 👍